### ISSUE TITLE
Cross entropy using bounds

### ISSUE THREAD

**Issue by mil-ad** [KaidiXu/auto_LiRPA#4]
So I've been staring at how the robustness term in the loss is computed and it's not immediately clear to me what's going on:

```py
lb_padded = torch.cat((torch.zeros(size=(lb.size(0), 1), dtype=lb.dtype, device=lb.device), lb), dim=1)
fake_labels = torch.zeros(size=(lb.size(0),), dtype=torch.int64, device=lb.device)
robust_ce = CrossEntropyLoss()(-lb_padded, fake_labels)
```
I'm probably just missing some nifty numpy trick but I can't follow above. For instance why isn't the upper bound part of this? Shouldn't you have a one in `fake_labels` at the padded index?

If I have a batch size of 100 and 10 classes, my `upper_bound` and `lower_bound` will be of shape `(100,10)` and my specification will be `(100, 9, 10)`. So in case of IBP I can construct worst-case logits by picking up the corresponding value from `lower_bound` for the correct class, and from `upper_bound` for incorrect classes. Do you want to construct a worst-case logit per specification (i.e. worst case logit has the shape `(100,9, 10)`? Or are you picking 9 logits from upper-bound and 1 from lower-bound to construct a single logit (per example) to form a worst-case logits with shape `(100, 10)`?

**Comment by shizhouxing** [comment]
Hi @mi-lad , 

Thanks for the question. For `lb_padded` here, we are computing the lower bound of the *margin* instead of the logits, since we have specification
```
c = torch.eye(num_class).type_as(data)[labels].unsqueeze(1) - torch.eye(num_class).type_as(data).unsqueeze(0)
```

You may refer to Appendix B.3 in our paper, and it gives 

L(f(X),y) = log sum_i exp(-g(X,y)_i) = -log exp(-g(X,y)_y)/(sum_i exp(-g(X,y)_i)),

where exp(-g(X,y)_y)=1. So, to compute the upper bound of L(f(X),y), we compute the lower bound of g(X,y)_i and then pass -g(X,y)_i, i.e., `-lb_padded`, to `CrossEntropyLoss`. Bounding margins directly should give tighter results than first computing a lower bound for the correct class and upper bounds for incorrect classes and then computing margins. 

For `fake_labels`, it means that we always put the margin between the real label and itself (i.e., zero) at first in `lb_padded`, so we always set `fake_labels` to 0 for computing `robust_ce` with `CrossEntropyLoss`.
